### PR TITLE
fix: Place permission middleware ahead of all method implementations

### LINKED ISSUE #24331: Stop implementing RPC methods ahead of the permission middleware
Ref: https://github.com/MetaMask/metamask-mobile/issues/9492

Some RPC methods, e.g. `wallet_switchEthereumChain` and `wallet_addEthereumChain`, are implemented ahead of the permission middleware. This makes it difficult to keep track of our restricted vs. unrestricted methods, while also literally circumventing our permission system. Instead, we should reorder our `json-rpc-engine` middleware stack such that every method hits the permission middleware, and all unrestricted methods have to be enumerated in [this list](https://github.com/MetaMask/metamask-extension/blob/2c457705f7288325c1ccbc26f117a07a32320684/app/scripts/controllers/permissions/specifications.js#L232-L290).

### PR DESCRIPTION
**Pull request by rekmarks**
## **Description**

Adds all unrestricted RPC methods to the `unrestrictedMethods` array passed to the permission controller, and moves the permission middleware ahead of all RPC method implementations in `setupProviderEngine`. This forces us to add all methods that aren't permission to the `unrestrictedMethods` array, as a "method not found" error will be thrown otherwise. It's safer this way.

Due the requirement that it swallows authorization errors, the implementation of `eth_accounts` is moved to a new "legacy method middleware", which we by necessity place ahead of the permission middleware. In addition, the responsibility of rejecting intentionally unsupported RPC methods has been moved to a dedicated middleware in front of the legacy method middleware.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24472?quickstart=1)

## **Related issues**

Fixes: #24331 

## **Manual testing steps**

1. Go the test dapp
2. Ensure that everything is working normally
3. Repeat these steps for Snaps and MMI

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by rekmarks** [comment]
@MetaMask/snaps-devs @MetaMask/mmi tagging you guys to confirm that your RPC methods continue to work as expected.

**Comment by rekmarks** [comment]
Converting to draft while figuring out how best to fix `wallet_snap` (i.e. the sugar for `wallet_invokeSnap`), whose implementation assumes that it runs before the permission middleware.

**Comment by rekmarks** [comment]
[test-e2e-chrome-mv3 appears to be consistently failing](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/82679/workflows/45e43639-4690-4259-8c15-2db4cdf4a280/jobs/2950731) for `snap-account-signatures.spec.ts`, but when I test using Flask MV3 and https://metamask.github.io/snap-simple-keyring/1.1.1/ on https://metamask.github.io/test-dapp/ everything appears to be in working order.

**Comment by danjm** [comment]
>test-e2e-chrome-mv3 [appears to be consistently failing](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/82679/workflows/45e43639-4690-4259-8c15-2db4cdf4a280/jobs/2950731) for snap-account-signatures.spec.ts, but when I try the flows manually using Flask MV3 and https://metamask.github.io/snap-simple-keyring/1.1.1/ on https://metamask.github.io/test-dapp/ everything appears to be in working order.

Those failures are fixed by https://github.com/MetaMask/metamask-extension/pull/24672

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -5040,13 +5036,19 @@ export default class MetamaskController extends EventEmitter {
...
 
-    // create filter polyfill middleware
-    const filterMiddleware = createFilterMiddleware(proxyClient);
+    // If the origin is not in the selectedNetworkController's `domains` state
+    // when the provider engine is created, the selectedNetworkController will
+    // fetch the globally selected networkClient from the networkController and wrap
+    // it in a proxy which can be switched to use its own state if/when the origin
+    // is added to the `domains` state
+    const proxyClient =
+      this.selectedNetworkController.getProviderAndBlockTracker(origin);
```
[the remark this answers is not in the document]
Comment (legobeat) [reply app/scripts/metamask-controller.js:5077]: The thing one could imagine would be if something further up has been relying on specific error being thrown from `getProviderAndBlockTracker` under certain conditions, but sounds far-fetched and move makes sense to me.

File: app/scripts/lib/rpc-method-middleware/createMethodMiddleware.js [hunk 2]
```diff
@@ -2,75 +2,92 @@ import { permissionRpcMethods } from '@metamask/permission-controller';
...
 
 const allHandlers = [...localHandlers, ...permissionRpcMethods.handlers];
 
-const handlerMap = allHandlers.reduce((map, handler) => {
-  for (const methodName of handler.methodNames) {
-    map[methodName] = handler;
-  }
-  return map;
-}, {});
+// The primary home of RPC method implementations in MetaMask. MUST succeed
```
Comment (adonesky1) [inline app/scripts/lib/rpc-method-middleware/createMethodMiddleware.js:9]: [nit] succeed may be a confusing word here
```suggestion
// The primary home of RPC method implementations in MetaMask. MUST follow
```

File: app/scripts/lib/rpc-method-middleware/handlers/index.ts [hunk 3]
```diff
@@ -35,4 +34,4 @@ const handlers = [
   ///: END:ONLY_INCLUDE_IF
 ];
 
-export default handlers;
+export const legacyHandlers = [ethAccounts];
```
Comment (adonesky1) [inline app/scripts/lib/rpc-method-middleware/handlers/index.ts:37]: not a huge deal but wondering if "legacy" is the right framing here? It seems like the main reason this is separated out is that you need to be able to call this method and receive an empty array if there is no authorization? Perhaps that is a legacy pattern in that we don't want to it to be reused but legacyHandlers implies to me that this is method that is set for deprecation or something like that

Comment (adonesky1) [reply app/scripts/lib/rpc-method-middleware/handlers/index.ts:37]: > Perhaps that is a legacy pattern in that we don't want to it to be reused

maybe this is the proper use of legacy... 🤔

**Review by adonesky1 — APPROVED** [review]
Looks good to me!

**Review by mcmire — APPROVED** [review]
Description of problem makes sense as does the solution. LGTM 👍🏻

**Pull request merged**